# ETL PROJECT - RETAIL TRANSACTIONS

The dataset it will be used in this project is from Kaggle plataform. It contains information about retail transactions conducted online. It contains information about customer purchases, including the invoice number, stock code, description of the items purchased, quantity, unit price, invoice date, customer ID, and country.


## 1.1 - Columns Dictionary



**1) InvoiceNo:** A unique identifier for each transaction or invoice.

**2) StockCode:** A code representing the stock or item purchased.

**3) Description:** A textual description of the item purchased.

**4) Quantity:** The quantity of the item purchased in each transaction.

**5) InvoiceDate:** The date and time when the transaction occurred.

**6) UnitPrice:** The price per unit of the item purchased.

**7) CustomerID:** The unique identifier for the customer making the purchase.

**8) Country:** The country where the transaction took place.


## 1- Extract


The table it will be used it was in Google Big Query plataform. So it will be necessary extract all the informations from there.

In [2]:
#import dependencies

import os
import pandas as pd
from google.cloud import bigquery
from google.oauth2 import service_account
import pandas_gbq


In [3]:
path = 'C:/Users/vanes/OneDrive/Documentos/PROJETOS_PORTFOLIO/ETL/'

In [7]:

# Set the path to your service account key JSON file
os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = path+"gbq.json"

# Your existing code
selectQuery = """SELECT * FROM etl-project-416319.project_1.customers_final"""
bigqueryClient = bigquery.Client()
df = bigqueryClient.query(selectQuery).to_dataframe()
df.to_csv("customer.csv", index=False)


In [8]:
df

InvoiceNo     StockCode                          Description Quantity  \
0         537032         21275                                    ?      -30   
1         538090         20956                                    ?     -723   
2         539494         21479                                    ?      752   
3         540100         22837                                    ?     -106   
4         540558         21258                                    ?      -29   
...          ...           ...                                  ...      ...   
541904    554550        47566B  incorrectly credited C550456 see 47     1300   
541905    554553        47566B  incorrectly credited C550456 see 47    -1300   
541906    547704         21621  rcvd be air temp fix for dotcom sit      100   
541907    556955  gift_0001_20  to push order througha s stock was        10   
541908    562113         22016  Dotcomgiftshop Gift Voucher £100.00        1   

                InvoiceDate UnitPrice CustomerID         Country  
0       2010-12-03 16:50:00       0.0        nan  United Kingdom  
1       2010-12-09 14:48:00       0.0        nan  United Kingdom  
2       2010-12-20 10:36:00       0.0        nan  United Kingdom  
3       2011-01-04 16:53:00       0.0        nan  United Kingdom  
4       2011-01-10 10:04:00       0.0        nan  United Kingdom  
...                     ...       ...        ...             ...  
541904  2011-05-25 09:57:00       0.0        nan  United Kingdom  
541905  2011-05-25 09:59:00       0.0        nan  United Kingdom  
541906  2011-03-24 17:51:00       0.0        nan  United Kingdom  
541907  2011-06-16 09:04:00       0.0        nan  United Kingdom  
541908  2011-08-02 16:11:00     83.33        nan  United Kingdom  

[541909 rows x 8 columns]

## 2- Transform

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   InvoiceNo    541909 non-null  object
 1   StockCode    541909 non-null  object
 2   Description  541909 non-null  object
 3   Quantity     541909 non-null  object
 4   InvoiceDate  541909 non-null  object
 5   UnitPrice    541909 non-null  object
 6   CustomerID   541909 non-null  object
 7   Country      541909 non-null  object
dtypes: object(8)
memory usage: 33.1+ MB


In [10]:
# checking for columns types

df.dtypes

InvoiceNo      object
StockCode      object
Description    object
Quantity       object
InvoiceDate    object
UnitPrice      object
CustomerID     object
Country        object
dtype: object

In [11]:
df.shape

(541909, 8)

In [12]:
# checking for duplicates

df.duplicated().sum()

5268

In [13]:

# pick up duplicated datas
duplicated = df[df.duplicated()]


print(duplicated)

# Se desejar contar o número de duplicados encontrados
print("Número de duplicados:", len(duplicated))


       InvoiceNo StockCode                          Description Quantity  \
2015      572344         M                               Manual       48   
2016      572344         M                               Manual       48   
2017      572344         M                               Manual       48   
2018      572344         M                               Manual       48   
2019      572344         M                               Manual       48   
...          ...       ...                                  ...      ...   
541306    558710    18097C  WHITE TALL PORCELAIN T-LIGHT HOLDER        1   
541308    563924    18097C  WHITE TALL PORCELAIN T-LIGHT HOLDER        1   
541309    563924    18097C  WHITE TALL PORCELAIN T-LIGHT HOLDER        1   
541326    577296    18097C  WHITE TALL PORCELAIN T-LIGHT HOLDER        1   
541481    574063     22577  WOODEN HEART CHRISTMAS SCANDINAVIAN        2   

                InvoiceDate UnitPrice CustomerID         Country  
2015    2011-10-24 1

In [14]:
# dropping duplicated data

df = df.drop_duplicates()
df

InvoiceNo     StockCode                          Description Quantity  \
0         537032         21275                                    ?      -30   
1         538090         20956                                    ?     -723   
2         539494         21479                                    ?      752   
3         540100         22837                                    ?     -106   
4         540558         21258                                    ?      -29   
...          ...           ...                                  ...      ...   
541904    554550        47566B  incorrectly credited C550456 see 47     1300   
541905    554553        47566B  incorrectly credited C550456 see 47    -1300   
541906    547704         21621  rcvd be air temp fix for dotcom sit      100   
541907    556955  gift_0001_20  to push order througha s stock was        10   
541908    562113         22016  Dotcomgiftshop Gift Voucher £100.00        1   

                InvoiceDate UnitPrice CustomerID         Country  
0       2010-12-03 16:50:00       0.0        nan  United Kingdom  
1       2010-12-09 14:48:00       0.0        nan  United Kingdom  
2       2010-12-20 10:36:00       0.0        nan  United Kingdom  
3       2011-01-04 16:53:00       0.0        nan  United Kingdom  
4       2011-01-10 10:04:00       0.0        nan  United Kingdom  
...                     ...       ...        ...             ...  
541904  2011-05-25 09:57:00       0.0        nan  United Kingdom  
541905  2011-05-25 09:59:00       0.0        nan  United Kingdom  
541906  2011-03-24 17:51:00       0.0        nan  United Kingdom  
541907  2011-06-16 09:04:00       0.0        nan  United Kingdom  
541908  2011-08-02 16:11:00     83.33        nan  United Kingdom  

[536641 rows x 8 columns]

In [21]:
# converting columns types

df.loc[:, 'Quantity'] = df['Quantity'].astype('int64')
df.loc[:, 'InvoceDate']= pd.to_datetime(df['InvoceDate'])
df.loc[:, 'UnitPrice'] = df['UnitPrice'].astype('float64')

In [22]:
# checking for missing values

df.isnull().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
InvoceDate     0
dtype: int64

In [23]:
df[df['Description'].isna()]

InvoiceNo     StockCode Description  Quantity  \
715       575506            C2        None       150   
859       547966           DOT        None      1000   
1589      540699          POST        None      1000   
1590      554857          POST        None       800   
1591      565556          POST        None       750   
...          ...           ...         ...       ...   
541774    542531      DCGS0072        None        -1   
541776    542532      DCGS0074        None        -1   
541824    561255     DCGS0066P        None        -3   
541875    564762  gift_0001_10        None        30   
541894    564761  gift_0001_30        None        30   

                     InvoiceDate  UnitPrice  CustomerID         Country  
715    2011-11-10 10:30:00+00:00        0.0         NaN  United Kingdom  
859    2011-03-28 15:49:00+00:00        0.0         NaN  United Kingdom  
1589   2011-01-11 09:32:00+00:00        0.0         NaN  United Kingdom  
1590   2011-05-27 10:08:00+00:00        0.0         NaN  United Kingdom  
1591   2011-09-05 12:14:00+00:00        0.0         NaN  United Kingdom  
...                          ...        ...         ...             ...  
541774 2011-01-28 13:08:00+00:00        0.0         NaN  United Kingdom  
541776 2011-01-28 13:09:00+00:00        0.0         NaN  United Kingdom  
541824 2011-07-26 11:52:00+00:00        0.0         NaN  United Kingdom  
541875 2011-08-30 10:48:00+00:00        0.0         NaN  United Kingdom  
541894 2011-08-30 10:48:00+00:00        0.0         NaN  United Kingdom  

[1454 rows x 8 columns]

In most rows there are missing values in the Description column, there are also missing values in CustomerID, and the UnitPrice is equal to 0. So those values are going to be deleted.

In [25]:
df = df.dropna(subset=['Description'])

In [23]:
df.dtypes

InvoiceNo              object
StockCode              object
Description            object
Quantity                int64
InvoiceDate            object
UnitPrice              object
CustomerID             object
Country                object
InvoceDate     datetime64[ns]
dtype: object

In [26]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description         0
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     133583
Country             0
dtype: int64

In [28]:
df_null_customer=df[df['CustomerID'].isna()]

In [29]:
df_null_customer.describe()

Quantity      UnitPrice  CustomerID
count   133583.0  133583.000000         0.0
mean    2.120479       8.166272         NaN
std    62.072857     152.747198         NaN
min      -9600.0  -11062.060000         NaN
25%          1.0       1.630000         NaN
50%          1.0       3.290000         NaN
75%          3.0       5.790000         NaN
max       4000.0   17836.460000         NaN

In [30]:
df.describe()


Quantity      UnitPrice     CustomerID
count    535187.0  535187.000000  401604.000000
mean     9.671593       4.645242   15281.160818
std    219.059056      97.364810    1714.006089
min      -80995.0  -11062.060000   12346.000000
25%           1.0       1.250000   13939.000000
50%           3.0       2.080000   15145.000000
75%          10.0       4.130000   16784.000000
max       80995.0   38970.000000   18287.000000

The dataframe has missing values in the CustomerID column. If the remaining corresponding data for these missing IDs suggested that the purchase did not exist, one approach to consider would be to delete these records. Alternatively, if the objective of the analysis is to understand the customer profile and how they shop on the site, then these records without identification would hinder the analysis, and deleting them would be the best approach. Here, as the objective is to understand some sales metrics, which will be subsequently addressed in the dashboard construction, these values will be replaced. In this column, the smallest ID is 12346 and the largest is 18287. Therefore, to facilitate the identification of these records that were null, they will be replaced by the value 1. Thus, whenever the value 1 appears in the CustomerID column, it means that it was a missing value.

In [32]:
df['CustomerID'].fillna(1, inplace=True)


In [33]:
# transform CustomerID column type into integer

df['CustomerID'] = df['CustomerID'].astype(int)

In [34]:
df

InvoiceNo     StockCode                         Description  Quantity  \
0        A563185             B                     Adjust bad debt         1   
1        A563186             B                     Adjust bad debt         1   
2        A563187             B                     Adjust bad debt         1   
3        C552650             D                            Discount       -18   
4        C545478             D                            Discount      -720   
...          ...           ...                                 ...       ...   
541904    561513  gift_0001_40  Dotcomgiftshop Gift Voucher £40.00         1   
541905    539958  gift_0001_50  Dotcomgiftshop Gift Voucher £50.00         1   
541906    552232  gift_0001_50  Dotcomgiftshop Gift Voucher £50.00         1   
541907    558066  gift_0001_50  Dotcomgiftshop Gift Voucher £50.00         1   
541908    558614  gift_0001_50  Dotcomgiftshop Gift Voucher £50.00         1   

                     InvoiceDate  UnitPrice  CustomerID         Country  
0      2011-08-12 14:50:00+00:00   11062.06           1  United Kingdom  
1      2011-08-12 14:51:00+00:00  -11062.06           1  United Kingdom  
2      2011-08-12 14:52:00+00:00  -11062.06           1  United Kingdom  
3      2011-05-10 14:03:00+00:00       1.45       16672  United Kingdom  
4      2011-03-03 11:08:00+00:00       0.01       16422  United Kingdom  
...                          ...        ...         ...             ...  
541904 2011-07-27 15:12:00+00:00      33.33           1  United Kingdom  
541905 2010-12-23 13:26:00+00:00      42.55           1  United Kingdom  
541906 2011-05-06 15:54:00+00:00      41.67           1  United Kingdom  
541907 2011-06-24 15:45:00+00:00      41.67           1  United Kingdom  
541908 2011-06-30 15:56:00+00:00      41.67           1  United Kingdom  

[535187 rows x 8 columns]

# ok